source/extensions/network/dns_resolver/cares/dns_impl.cc

```diff
@@ -89,11 +115,23 @@ void DnsResolverImpl::initializeChannel(ares_options* options, int optmask) {
     static_cast<DnsResolverImpl*>(arg)->onAresSocketStateChange(fd, read, write);
   };
   options->sock_state_cb_data = this;
-  ares_init_options(&channel_, options, optmask | ARES_OPT_SOCK_STATE_CB);
+  optmask |= ARES_OPT_SOCK_STATE_CB;
+
+  // Point channel to fallback name servers if they exist.
+  if (user_defined_resolvers_.has_value() &&
+      absl::holds_alternative<std::vector<in_addr>>(user_defined_resolvers_.value())) {
+    auto& resolvers_vector = absl::get<std::vector<in_addr>>(user_defined_resolvers_.value());
+    options->nservers = resolvers_vector.size();
+    options->servers = &resolvers_vector.front();
+    optmask |= ARES_OPT_SERVERS;
+  }
+  ares_init_options(&channel_, options, optmask);
 
-  // Ensure that the channel points to custom resolvers, if they exist.
-  if (resolvers_csv_.has_value()) {
-    int result = ares_set_servers_ports_csv(channel_, resolvers_csv_->c_str());
+  // Otherwise point channel to override name servers if they exist.
```

Comment:
This is still really confusing. Why are we using the csv for one and the init options for the other? Can you add more comments about what the difference is if there is a difference?

Reply:
Added comments. Ultimately, the way we are exposing fallback semantics on Envoy is not really an official c-ares API as much as it is co-opting c-ares' implementation. So the format c-ares expects for the addresses in the options struct is different (and more restrictive ipv4 only, no ports) than the one on the official `set_ports` APIs. Hence the variant with different types.

I added comments to explain that without inundating with details about the c-ares impl.

Reply:
Actually. This might be a moot point due to a [test a just wrote](https://github.com/envoyproxy/envoy/pull/19010/files#diff-0ccaae5ed50ce7c8e5cceaf7a193d8ce4dcbe664d036aadefc72b97a9fad1cb4R430). Investigating...